helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -539,7 +539,15 @@ private void subscribeChildChange(String path, NotificationContext.Type callback
         logger.debug(_manager.getInstanceName() + " subscribes child-change. path: " + path
             + ", listener: " + _listener);
       }
-      _zkClient.subscribeChildChanges(path, this);
+      // In the lifecycle of CallbackHandler, INIT is the first stage of registration of watch.
+      // For some usage case such as current state, the path can be created later. Thus we would
+      // install watch anyway event the path is not yet created.
+      // Later, CALLBACK type, the CallbackHandler already registered the watch and knows the
+      // path was created. Here, to avoid leaking path in ZooKeeper server, we would not let
+      // CallbackHandler to install exists watch, namely watch for path not existing.
+      // Note when path is removed, the CallbackHanler would remove itself from ZkHelixManager too
+      // to avoid leaking a CallbackHandler.
+      _zkClient.subscribeChildChanges(path, this, callbackType != Type.INIT);
```

Comment:
Why not check the result here? If subscribe fails, shall we just ignore it?

Reply:
If CallbackType is CALLBACK, ignore it is just want we want. That is exactly the case of async removal of a task current state node from server would do. 

If CallbackType is INIT, it would not fail. 

So let us add a debug log?

Reply:
Agreed